go-controller/pkg/ovn/topology_version.go

```diff
@@ -0,0 +1,142 @@
+	ovntypes "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1apply "k8s.io/client-go/applyconfigurations/core/v1"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+)
+
+func (oc *Controller) ovnTopologyCleanup(ctx context.Context) error {
+	ver, err := oc.determineOVNTopoVersionFromOVN(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Cleanup address sets in non dual stack formats in all versions known to possibly exist.
+	if ver <= ovntypes.OvnPortBindingTopoVersion {
+		err = addressset.NonDualStackAddressSetCleanup(oc.nbClient)
+	}
+	return err
+}
+
+// reportTopologyVersion saves the topology version to two places:
+// - an ExternalID on the GatewayRouter in nbdb
+// - a ConfigMap. This is used by nodes to determine the cluster's topology
+func (oc *Controller) reportTopologyVersion(ctx context.Context) error {
+	currentTopologyVersion := strconv.Itoa(ovntypes.OvnCurrentTopologyVersion)
+	logicalRouterRes := []nbdb.LogicalRouter{}
+	ctx, cancel := context.WithTimeout(ctx, ovntypes.OVSDBTimeout)
+	defer cancel()
+	if err := oc.nbClient.WhereCache(func(lr *nbdb.LogicalRouter) bool {
+		return lr.Name == ovntypes.OVNClusterRouter
+	}).List(ctx, &logicalRouterRes); err != nil {
+		return fmt.Errorf("failed in retrieving %s, error: %v", ovntypes.OVNClusterRouter, err)
+	}
+	// Update topology version on distributed cluster router
+	logicalRouterRes[0].ExternalIDs["k8s-ovn-topo-version"] = currentTopologyVersion
+	logicalRouter := nbdb.LogicalRouter{
+		Name:        ovntypes.OVNClusterRouter,
+		ExternalIDs: logicalRouterRes[0].ExternalIDs,
+	}
+	opModel := libovsdbops.OperationModel{
+		Name:           logicalRouter.Name,
+		Model:          &logicalRouter,
+		ModelPredicate: func(lr *nbdb.LogicalRouter) bool { return lr.Name == ovntypes.OVNClusterRouter },
+		OnModelUpdates: []interface{}{
+			&logicalRouter.ExternalIDs,
+		},
+		ErrNotFound: true,
+	}
+	if _, err := oc.modelClient.CreateOrUpdate(opModel); err != nil {
+		return fmt.Errorf("failed to generate set topology version in OVN, err: %v", err)
+	}
+	klog.Infof("Updated GatewayRouter topology version to %s", currentTopologyVersion)
+
+	// Report topology version in a ConfigMap
+	// (we used to report this via annotations on our Node)
+	cm := corev1apply.ConfigMap(ovntypes.OvnK8sStatusCMName, globalconfig.Kubernetes.OVNConfigNamespace)
+	cm.Data = map[string]string{ovntypes.OvnK8sStatusKeyTopoVersion: currentTopologyVersion}
+	if _, err := oc.client.CoreV1().ConfigMaps(globalconfig.Kubernetes.OVNConfigNamespace).Apply(ctx, cm, metav1.ApplyOptions{
+		Force:        true,
+		FieldManager: "ovn-kubernetes",
+	}); err != nil {
+		return err
+	}
+
+	klog.Infof("Updated ConfigMap %s/%s topology version to %s", *cm.Namespace, *cm.Name, currentTopologyVersion)
+
+	// Unset the old topology annotation on all Node objects
```

Comment:
nit: maybe modularize this part into separate function that can be called from here to stop overloading the "reportTopologyVersion"

Reply:
sure